goldmane/pkg/daemon/config.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2025 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/calico/goldmane/pkg/aggregator"
+	"github.com/projectcalico/calico/goldmane/pkg/aggregator/bucketing"
+	"github.com/projectcalico/calico/goldmane/pkg/internal/utils"
+)
+
+type goldmaneFileConfig struct {
+	EmitFlows bool `json:"emitFlows"`
+}
+
+type sinkManager struct {
+	aggregator *aggregator.LogAggregator
+	sink       bucketing.Sink
+	upd        chan struct{}
+	watchFn    func(context.Context)
+	path       string
+}
+
+func newSinkManager(agg *aggregator.LogAggregator, sink bucketing.Sink, path string) (*sinkManager, error) {
+	onUpdate := make(chan struct{}, 1)
+
+	// Watch for changes to the input file.
+	watchFn, err := utils.WatchFilesFn(onUpdate, path)
+	if err != nil {
+		return nil, err
+	}
+
+	if sink == nil {
+		return nil, fmt.Errorf("a sink must be provided")
+	}
+	if agg == nil {
+		return nil, fmt.Errorf("an aggregator must be provided")
+	}
+
+	e := sinkManager{
+		upd:        onUpdate,
+		watchFn:    watchFn,
+		aggregator: agg,
+		sink:       sink,
+		path:       path,
+	}
+	return &e, nil
+}
+
+func (f *sinkManager) run(ctx context.Context) {
+	logrus.WithField("path", f.path).Info("Starting sink manager with config path")
+	defer logrus.Warn("Sink manager exiting")
+	defer close(f.upd)
+
+	// Start of day - check if we should enable the sink.
+	if sinkEnabled(f.path) {
+		logrus.Debug("Sink enabled at startup")
+		f.aggregator.SetSink(f.sink)
+	}
+	logrus.Info("Sink manager started")
+
+	// Start the file watch.
+	go f.watchFn(ctx)
+
+	go func() {
+		// If enablement changes, update the sink.
+		for range f.upd {
```

Comment:
I think given that each update here would trigger a call to emit the flows it would be worth to ensure we're not getting duplicate or unnecessary updates, i.e. ensuring that updates are only received if the file contents have changed.

You could also just track if it's already enabled or not, but this is the second place where we ideally don't want unecessary updates so fixing the issue at the source seems pragmatic.

Reply:
The call to emit flows is already de-duplicated, so if there's nothing new to do then no flows will be emitted and minimal cycles will be used. I don't think there's a lot of value de-duplicating here - it will just introduce opportunities for us to get something wrong.

The code that checks if flows need to be emitted already needs to be capable of deduplicating regardless of how this loop behaves.

Reply:
I can't say I really understand why you'd want the file watcher to send events that don't need to be emitted because nothing has actually changed, especially given that you're already writing de duplication logic outside the watcher (i.e. in the emitter/http_client) where you could just make this inherent in it.

This just seems like requiring any user of the file watcher to fix a flaw with it.

Anyway, I'll leave it at that.